### PR TITLE
WS tests - unification & transpilation

### PR DESCRIPTION
**Pull request by ttodua**

### CODE REVIEW

File: ts/src/pro/test/.eslintrc [hunk 1]
```diff
@@ -1,20 +1,19 @@
 {
   "env": {
-    "node": true,
-    "mocha": true
+      "node": true
```
Comment (ttodua) [inline ts/src/pro/test/.eslintrc:3]: just placed the same eslintrc file from rest test's dir in WS folder

**Comment by ttodua** [comment]
three ws .ts individual test remains to be recoded. will do tomorrow.

**Comment by ttodua** [comment]
note about changes:
- removed run-tests-ws.js, as everything is now integrated in `run-tests.js`, just running it with `--ws` flag triggers ws tests.
- removed language-specific `test ks|py|php` files from php/py pro directiories, and instead they will be tested through main ("rest") `test.py|ts|php` files.
- moved transpilation of tests (including WS tests) from `transpileWs.js` into `transpile.js` (this step was needed in order to simplify things and avoid complexity)
- removed all individual ws-test files from py/php folders.
- added `custom` folders in js/py/php, which are meant to contain only language-specific scripts & tests, and are not transpilable. all other files are transpilable from `pro/test` directory & subdirectories

File: python/ccxt/async_support/base/ws/cache.py [hunk 2]
```diff
@@ -47,6 +47,10 @@ def __getitem__(self, item):
         else:
             return deque[item]
 
+    def __getattr__(self, attr_name):
+        # support snake_case
+        if (attr_name == 'get_limit'):
+            return self.getLimit
```
Comment (ttodua) [inline python/ccxt/async_support/base/ws/cache.py:54]: ast transpiler causes the `getLimit` method to be transpiled int snake-case, so this was easiest way to fix

File: python/ccxt/pro/test/base/test_cache.py [hunk 3]
```diff
@@ -0,0 +1,914 @@
+import os
```
Comment (carlosmiei) [inline python/ccxt/pro/test/base/test_cache.py:1]: @ttodua test_Cache/orderbook were only moved right?

File: build/transpileWS.js [hunk 4]
```diff
@@ -147,81 +147,6 @@ class CCXTProTranspiler extends Transpiler {
...
-        const php = this.getPHPPreamble () + phpHeader + phpBody
-
-        log.magenta ('→', pyFile.yellow)
-        log.magenta ('→', phpFile.yellow)
-
-        overwriteFile (pyFile, python)
-        overwriteFile (phpFile, php)
-    }
-
-    // ------------------------------------------------------------------------
```
Comment (carlosmiei) [inline build/transpileWS.js:223]: why was this removed?

File: python/ccxt/async_support/base/ws/cache.py (hunk 2 above)
Comment (carlosmiei) [reply python/ccxt/async_support/base/ws/cache.py:54]: we can do this or use a regex

File: ts/src/pro/test/Exchange/test.watchBalance.ts [hunk 5]
```diff
@@ -1,99 +1,23 @@
...
-        return;
-    }
-
-    console.log ('watching balance...');
-
-    let now = Date.now ();
-    const ends = now + 10000;
-
+    let now = exchange.milliseconds ();
+    const ends = now + exchange.wsMethodsTestTimeoutMs;
```
Comment (carlosmiei) [inline ts/src/pro/test/Exchange/test.watchBalance.ts:8]: why use a tests-only property in the exchange?

**Comment by carlosmiei** [comment]
@ttodua can you check if we can enable many of the disabled ws tests we have in skip-tests?

File: python/ccxt/pro/test/base/test_cache.py (hunk 3 above)
Comment (ttodua) [reply python/ccxt/pro/test/base/test_cache.py:1]: yes, just moved into folder

File: build/transpileWS.js (hunk 4 above)
Comment (ttodua) [reply build/transpileWS.js:223]: all test files are being transpiled from transpile.js from now. we will have many other test files in the future, so listing them one by one and dedicating them separate methods is a 'no-go' imo.

File: ts/src/pro/test/Exchange/test.watchBalance.ts (hunk 5 above)
Comment (ttodua) [reply ts/src/pro/test/Exchange/test.watchBalance.ts:8]: here it is being set: https://github.com/ccxt/ccxt/blob/a81f8cc8a1a672f4d0522e6e3cc52a26f0b82d68/ts/src/test/test.ts#L312
this was to avoid intrusive changes to method signatures to pass that value as argument, so setting as exchange prop seems a possible way for now. wdyt?

**Comment by ttodua** [comment]
> @ttodua can you check if we can enable many of the disabled ws tests we have in skip-tests?

i'll hadn't, but i will asap.

**Comment by ttodua** [comment]
@carlosmiei I've removed almost all WS skips and seems many exchanges would pass them

**Comment by ttodua** [comment]
@carlosmiei yes, WS tests passed for all exchanges in travis. I have just to add WS prefix/suffix in logs to differentiate them

File: ts/src/pro/test/Exchange/test.watchBalance.ts (hunk 5 above)
Comment (carlosmiei) [reply ts/src/pro/test/Exchange/test.watchBalance.ts:8]: I think I would rather pass it explicitly in the method's signature, it's easier to understand where it is coming from. 

Second question, I'm not sure this is the "timeout" control we need, what if the method never returns inside the  loop? it will hang "forever" right?

File: build/transpileWS.js (hunk 4 above)
Comment (carlosmiei) [reply build/transpileWS.js:223]: The thing is I'm not sure we should mix these tests transpiling with the normal tests, even in the c# side I kept them separated. each test type has different needs and keeping everything within the same method with a lot of (if isWsCacheTest) etc gets confusing very quickly

Comment (ttodua) [reply build/transpileWS.js:223]: but as we use Ast transpiler, it doesn't have much difference whether the transpiling file is from 'examples' folder, from 'tests', from 'rest' exchagnes or from 'ws' exchanges. if we separate the logic, then we shuold have `transpile-examples, transpile-tests, transpile-rest, transpile-ws' methods, and on top of that, `transpile-cache, transpile-orderbook, transpile-xyz...` which is more hard to maintain and track. when doing this from same `transpileTests` method, it's easier to control and manage everything

Comment (ttodua) [reply build/transpileWS.js:223]: so, it actually does not matter for transpiler whether the file being transpiled is from ws folder or from elsewhere.
but if you say otherwise, i will try to tear down them and migrate ws parts...

Comment (carlosmiei) [reply build/transpileWS.js:223]: we can do everything every from transpile tests but I would keep it in different methods like we have now, like
```
transpileCacheTest()
transpileOrderBookTest()
```
In the future if we need to apply a specific fix it will be much easier

File: ts/src/pro/test/Exchange/test.watchBalance.ts (hunk 5 above)
Comment (carlosmiei) [reply ts/src/pro/test/Exchange/test.watchBalance.ts:8]: 1) why does it take many changes? isn't just adding a new parameter to each method?
for instance in c# I would have to explicitly define this property which is a bit odd since this is tests-only

Comment (ttodua) [reply ts/src/pro/test/Exchange/test.watchBalance.ts:8]: we could add that in `options['wsTimeout']` etc.. what would avoid defining it in c#

Comment (ttodua) [reply ts/src/pro/test/Exchange/test.watchBalance.ts:8]: or I can leave that hardcoded as it was before?

Comment (carlosmiei) [reply ts/src/pro/test/Exchange/test.watchBalance.ts:8]: @ttodua Question, how many changes would it require to change this
```
testWatchOrderBook (exchange, skippedProperties, symbol)
```

to this 
```
testWatchOrderBook (exchange, skippedProperties, symbol, timeOut = 20000)
```

Comment (ttodua) [reply ts/src/pro/test/Exchange/test.watchBalance.ts:8]: I am also making some other updates in transpiler.js, so you might give up this PR review for a moment, until i turn it from **draft** into ready

**Comment by ttodua** [comment]
@car

File: build/transpile.js [hunk 6]
```diff
@@ -2171,12 +2175,13 @@ class Transpiler {
...
             phpBase: './php/test/base/',
         };
-
+       
         let baseTests = fs.readdirSync (baseFolders.tsBase).filter(filename => filename.endsWith('.ts')).map(filename => filename.replace('.ts', ''));
         const exchangeTests = fs.readdirSync (baseFolders.ts).filter(filename => filename.endsWith('.ts')).map(filename => filename.replace('.ts', ''));
 
         // ignore throttle test for now
         baseTests = baseTests.filter (filename => filename !== 'test.throttle');
+        this.createBaseInitFile(baseFolders.pyBase, baseTests);
```
Comment (ttodua) [inline build/transpile.js:2184]: just moved from 20 lines below to here

**Comment by carlosmiei** [comment]
@ttodua I think something might be off, we have some WS failures but the output don't show up
image

File: build/transpileWS.js (hunk 4 above)
Comment (carlosmiei) [reply build/transpileWS.js:229]: @ttodua can you please keep the context-specific tests (orderbook, cache, etc) in their isolated methods, as they were before? As I said I don't think that mixing them is a good idea just makes thing more confusing

Comment (ttodua) [reply build/transpileWS.js:229]: ok, i'll try to revert.

Comment (carlosmiei) [reply build/transpileWS.js:229]: But they're fundamentally different tests and at any time we might need to apply specific fixes to them

Comment (carlosmiei) [reply build/transpileWS.js:229]: if everything is in a single place it will be much harder

Comment (ttodua) [reply build/transpileWS.js:229]: the thing is that the code is being transpiled well and any update would not cause much need to do language-specific complex edits (because we have unified tests & checks almost for everyting, and still increasing details & coverage), so, i was forwarding to avoid separate methods like "transpile_test_file_A, transpile_test_file_B," and etc.. 
but as you say, i'll do so ✌️

File: ts/src/pro/test/test.ts
Comment (ttodua) [inline ts/src/pro/test/test.ts:1]: no need any test.ts/py/php files in `pro` folder anymore.

File: php/pro/BaseCache.php [hunk 7]
```diff
@@ -53,4 +53,19 @@ public function offsetUnset($index) : void {
...
+    }
+
+    // tmp support for type bcz of transpiler snake_case
+    public function __call($name, $arguments)
+    {
+        $camelCased = $this->convert_from_snake_to_camel_case($name);
+        if (method_exists($this, $camelCased)) {
+            return call_user_func_array(array($this, $camelCased), $arguments);
+        }
+        trigger_error('Call to undefined method '.__CLASS__.'::'.$name.'()', E_USER_ERROR);
```
Comment (carlosmiei) [inline php/pro/BaseCache.php:69]: @ttodua can't we just add 

```
    public function get_limit() : int {
        return $this->getLimit()
    }
```
instead?

File: python/ccxt/async_support/base/ws/cache.py (hunk 2 above)
Comment (carlosmiei) [reply python/ccxt/async_support/base/ws/cache.py:54]: same question here, why not `def get_limit` explicit is better

File: php/pro/BaseCache.php (hunk 7 above)
Comment (ttodua) [reply php/pro/BaseCache.php:69]: I though automatic way would have been better, because anytime in future there might be amended function-names, added functions or renamed things. why to hardcode?

Comment (ttodua) [reply php/pro/BaseCache.php:69]: but ok i'll do that as you said

File: ts/src/pro/test/Exchange/test.watchOrderBook.ts [hunk 8]
```diff
@@ -1,65 +1,51 @@
...
+        assert (typeof response === 'object', exchange.id + ' ' + method + ' ' + symbol + ' must return an object. ' + exchange.json (response));
+        now = exchange.milliseconds ();
+        testOrderBook (exchange, skippedProperties, method, response, symbol);
     }
-
-    return response;
-};
+}
+
+function fixPhpObjectArray (exchange, response, skippedProperties) {
```
Comment (carlosmiei) [inline ts/src/pro/test/Exchange/test.watchOrderBook.ts:28]: I still don't like this :/, first does not replicate the behavior on the user side, second is not concurrent safe, third is PHP only. 

Can't we comment out the failing assertion in PHP for now, instead of adding this?

File: php/pro/BaseCache.php (hunk 7 above)
Comment (carlosmiei) [reply php/pro/BaseCache.php:69]: > be amended function-names, added functions or renamed things. why to hardcode
because if the transpiler transpiles getLimit to something weird is much easier to spot and besides that it simplifies the code

File: ts/src/pro/test/Exchange/test.watchOrderBook.ts (hunk 8 above)
Comment (carlosmiei) [reply ts/src/pro/test/Exchange/test.watchOrderBook.ts:28]: We remove this entirely, and do a custom regex replace to replace the failing assertion in PHP only,  no?

File: ts/src/pro/test/Exchange/test.watchOrderBook.ts [hunk 9]
```diff
@@ -1,65 +1,51 @@
...
+    exchange.setExchangeProperty ('quoteJsonNumbers', false);
+    const result = exchange.parseJson (exchange.json (response));
+    exchange.setExchangeProperty ('quoteJsonNumbers', existingJqMode);
+    // temporary fix, because after json.strinfigy->parse, 'undefined' members are removed
+    skippedProperties['timestamp'] = true;
+    skippedProperties['datetime'] = true;
+    skippedProperties['nonce'] = true;
+    // ### temporarily fix some bugs for PHP (before they are fixed in library) ###
+    // 1) entries are being unordered in some cases, so before that separate issue is fixed, temporarily fix it here. for example, some entries are weird, like [[price, amount], [price, amount], ["1", amount]]]
+    result['asks'] = exchange.sortBy(result['asks'], 0, false);
```
Comment (carlosmiei) [inline ts/src/pro/test/Exchange/test.watchOrderBook.ts:40]: also, is your linter working properly? can see some errors there
image

File: ts/src/pro/test/Exchange/test.watchOrderBook.ts (hunk 8 above)
Comment (ttodua) [reply ts/src/pro/test/Exchange/test.watchOrderBook.ts:28]: Ok, i'll do it as to skip those orderbook tests in PHP temporarily

File: ts/src/pro/test/Exchange/test.watchOrderBook.ts (hunk 9 above)
Comment (ttodua) [reply ts/src/pro/test/Exchange/test.watchOrderBook.ts]: woah, good shot. i missed them because my linter didnt show anything - https://imgsh.net/i/l4x1cNM.png
hmm, strange, why. 

thanks for not, i'll look into that

File: python/ccxt/pro/test/test_order_book.py [hunk 10]
```diff
@@ -1,396 +0,0 @@
...
-
-# PLEASE DO NOT EDIT THIS FILE, IT IS GENERATED AND WILL BE OVERWRITTEN:
-# https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-contribute-code
-
-# ----------------------------------------------------------------------------
-
-from ccxt.async_support.base.ws.order_book import OrderBook, IndexedOrderBook, CountedOrderBook  # noqa: F402
-
-
-def equals(a, b):
```
Comment (carlosmiei) [inline python/ccxt/pro/test/test_order_book.py:17]: on the newest version we get this version of equals, which does not seem right 

image

Comment (carlosmiei) [reply python/ccxt/pro/test/test_order_book.py:17]: same thing in php

Comment (ttodua) [reply python/ccxt/pro/test/test_order_book.py:17]: can you plz elaborate on those functions? they have been there before and i dont actually know what they do there (havent touched and havent checked till date).
what should I do to them?

Comment (ttodua) [reply python/ccxt/pro/test/test_order_book.py:17]: ah wait, len(a) len(b) ?? ogrhhh

Comment (ttodua) [reply python/ccxt/pro/test/test_order_book.py:17]: fixed in both php & py

Comment (ttodua) [reply python/ccxt/pro/test/test_order_book.py:17]: if any other notes, lmk

File: ts/src/pro/test/Exchange/test.watchTrades.ts [hunk 11]
```diff
@@ -1,77 +1,30 @@
...
-import errors from '../../../base/errors.js';
-
-/*  ------------------------------------------------------------------------ */
-
-export default async (exchange, symbol) => {
-
-    // log (symbol.green, 'watching trades...')
+import testSharedMethods from '../../../test/Exchange/base/test.sharedMethods.js';
 
+async function testWatchTrades (exchange, skippedProperties, symbol) {
```
Comment (carlosmiei) [inline ts/src/pro/test/Exchange/test.watchTrades.ts:6]: @ttodua maybe we need to adjust the eslintrc?
image

Comment (carlosmiei) [reply ts/src/pro/test/Exchange/test.watchTrades.ts:6]: will fix this soon, found the reason

Comment (ttodua) [reply ts/src/pro/test/Exchange/test.watchTrades.ts:6]: ok, lmk me too what's the reason (i don't see those eslint warning, even though we interact with same eslintrc)

Comment (carlosmiei) [reply ts/src/pro/test/Exchange/test.watchTrades.ts:6]: you sure your eslint is working properly? can you run npm i again? we're missing a rule to disbale the jsdocs thing

File: skip-tests.json [hunk 12]
```diff
@@ -49,9 +51,7 @@
         }
     },
     "binance": {
```
Comment (carlosmiei) [inline skip-tests.json:53]: @ttodua u sure the php proxy is working properly? 
image
